Lucy numbering br

LuxxxLucy opened the pull request:
Regarding Issue #3882,

I mainly took references from [Greek Number Converter](https://www.russellcottrell.com/greek/utilities/GreekNumberConverter.htm) and [Greek Number](https://www.unco.edu/fraternity-sorority/resources/greek-alphabet.aspx)

I am for now using the lowercase and upper case Alpha as the symbol.

Some test cases (unicode level) are added in `numbering.rs`
```
    greek_number_tests! {
...
        long_complex_0: (97_554, "αΜθʹ, ͵ζφνδ", Case::Lower),
        long_complex_1: (2_056_839_184, "βΜκʹ, αΜ͵εχπγ, ͵θρπδ", Case::Lower),
        long_complex_2: (12_312_398_676, "βΜρκγʹ, αΜ͵ασλθ, ͵ηχοϛ", Case::Lower),
...
    }
```

Png rendered 
![image](https://github.com/typst/typst/assets/19356905/17a1095b-af11-4924-b871-6b68f66139dd)

JamesxX commented:
Would it not make more sense for the numbering to cover the entire alphabet rather than the first 10 characters? Base 26 (or however many characters greek has) rather than base 10

LuxxxLucy commented:
> Would it not make more sense for the numbering to cover the entire alphabet rather than the first 10 characters? Base 26 (or however many characters greek has) rather than base 10

Hi @JamesxX I think the main issue is definition on what is the Greek numeral system. The thing you just mentioned is called [lower-greek](https://w3c.github.io/predefined-counter-styles/#lower-greek) which represent 11 as λ (the 11-th char of the alphabet), while the greek numbering system is a decimal-like system (though with a different set of  chars), 11 is represented by "ιαʹ" (where "ι" is 10 and "α" is 1).

it seems to me that both should be implemented if permited.

laurmaedje commented on tests/suite/model/numbering.typ:
```diff
@@ -24,6 +24,25 @@
   [ for #i \ ]
 }
 
+--- numbering-greek ---
```
New numbering tests should not use reference images anymore, but rather assertions. The old ones should also be updated some time.

LuxxxLucy replied:
Update the test case, now only assertions, not sure whether it is what you want it to be. If so, do you want all the numbering test to be change to assertions in the numbering.type file like this?

(I previously thought that the typ and ref png should be the test on rendering results, and because of so for the numbering I added some local test to `numbering.rs` where the function is defined.)

laurmaedje replied:
>  If so, do you want all the numbering test to be change to assertions in the numbering.type file like this?

In principle yes, but that can be done separately. But open a PR if you'd like!

> I previously thought that the typ and ref png should be the test on rendering results, and because of so for the numbering I added some local test to numbering.rs where the function is defined.

I think the tests should just be where it's most convenient. If it's easier to test in Typst code, test it there, if it's simpler to test in Rust, test it there. This applies to the logic itself. The interface must be tested in Typst code to ensure that it works in integration.

laurmaedje commented on crates/typst/src/model/numbering.rs:
```diff
@@ -208,8 +208,11 @@ impl FromStr for NumberingPattern {
             let prefix = pattern[handled..i].into();
-            let case =
-                if c.is_uppercase() || c == '壹' { Case::Upper } else { Case::Lower };
+            let case = if c.is_uppercase() || c == '壹' || c == 'Α' {
```
why `c == 'Α'`? isn't `c.is_uppercase()` true for it?

probably we need to update `c.to_ascii_lowercase()` above since we have more than ASCII here

LuxxxLucy replied:
> why `c == 'Α'`? isn't `c.is_uppercase()` true for it?

Oh that is true; we do not need to do this. Updated.

> probably we need to update `c.to_ascii_lowercase()` above since we have more than ASCII here

Changed to use general to_lowercase()

laurmaedje reviewed (commented):
Here are a few comments, but not really relevant ones, since I can't judge the correctness of the Greek numbering itself. And then there is of course also still the question of the letter etc.

I'm not sure how best to proceed here.

laurmaedje commented:
I merged with main and tidied up the tests. While doing that I noticed that the number 2000000001 comes out differently with your code than with the [online converter](https://www.russellcottrell.com/greek/utilities/GreekNumberConverter.htm):
- Your code: "βΜκʹ, αΜαʹ
- Online converter: βΜκʹ, αʹ

laurmaedje commented:
Regarding the letter. It think we should use `α` and `Α`. Yes, the uppercase variant looks just like a normal A, but it is the only consistent choice. You can always write `sym.Alpha` and I also like the idea (forget where it was presented) to allow long-form numberings like `{greek}` and as soon as there is a brace-style numbering, implicit numbering is disabled.

LuxxxLucy commented:
> I merged with main and tidied up the tests. While doing that I noticed that the number 2000000001 comes out differently with your code than with the [online converter](https://www.russellcottrell.com/greek/utilities/GreekNumberConverter.htm):
> 
> * Your code: "βΜκʹ, αΜαʹ
> * Online converter: βΜκʹ, αʹ

Sorry, I did not realize it is so long ago and almost forgot this PR. I have just updated the PR with a fix and added more test cases.

> Regarding the letter. It think we should use `α` and `Α`.

the current way should be this case, using the `α` and `Α`.

>  I also like the idea (forget where it was presented) to allow long-form numberings like {greek} 

I like this idea as well!  I am interested it supporting it but do not know how to start with it, do you have some hint/high level ideas on how to do it? that would be great.

laurmaedje commented:
> I have just updated the PR with a fix and added more test cases.

Thanks!

> the current way should be this case, using the α and Α.

Yep.

> I like this idea as well! I am interested it supporting it but do not know how to start with it, do you have some hint/high level ideas on how to do it? that would be great.

Responded here: https://github.com/typst/typst/issues/1177#issuecomment-2451572160

laurmaedje commented:
Thank you!
